tools/bigquery-zos-mainframe-connector/gszutil/grecv/README.md

```diff
@@ -0,0 +1,90 @@
+# grecv
+
+grecv offloads decoding and encoding from the mainframe.
+
+The server is a java application which terminates TLS and decompresses a GZIP stream 
+of blocks sent by the client. The server maintains an array of decoders to 
+decode fields in the input dataset. After decoding, an ORC Writer appends batches 
+of data to an ORC file in GCS. The target ORC partition size is 128 MB. After a 
+partition exceeds the target size, a new writer is created automatically.
+
+The client reads blocks from the input dataset and sends round-robin across one or 
+more sockets to the server.
+
+
+## Protocol
+
+### Client Handshake
+
+Upon connecting, the client sends an integer indicating the number of 
+bytes in the serialized request message.
+The serialized request message is sent immediately after the length.
+The server replies with 200 if it was able to deserialize the message.
+The client then closes the connection.
+
+
+### Data upload
+
+The client connects one or more sockets and initializes a GZIPOutputStream for each.
+The server accepts exactly the number of connections specified during the handshake.
+Before each data block, the client sends an integer indicating the size of the block.
+A value of less than 1 is sent to indicate end of stream.
+Upon receiving the end of stream message, the server sends the number of blocks it has 
+received and closes the connection.
+The client confirms the block count and closes the connection. 
+
+
+## Deployment
+
+GCS is used as an artifact repository.
+Each version of grecv will have the following objects:
+- `mainframe-connector-version.jar` Application
+- `lib.tar` Application dependencies
+- `run.sh` Application launcher
+- `jre.tar` JVM
+- `pki.tar` TLS certificate and private key
+
+The `pkgUri` passed to grecv will be different for each environment.
```

Comment:
pkgUri was added to launch a dedicated instance of this service for each job, but this ultimately wasn't used, I think we should remove this and all related documentation. The default deployment should be to long-running VM or container.